UnmarshalKey implemented using pure reflection

dustmop opened the pull request:
### What does this PR do?

This version of UnmarshalKey doesn't rely upon the config having an internal `map[string]interface{}` to store data. Instead, it is implemented using reflection only and demonstrates the a node-based view of the config's data model.

### Motivation

Preparation for the new config model

### Additional Notes

### Possible Drawbacks / Trade-offs

### Describe how to test/QA your changes

Functionality is covered by unit tests

dustmop committed:
UnmarshalKey implemented using pure reflection

This version of UnmarshalKey doesn't rely upon the config having an internal map[string]interface{} to store data. Instead, it is implemented using reflection only and demonstrates the a node-based view of the config's data model.

dustmop committed:
Remove unused field to fix lint error

dustmop committed:
Allow boolean string to be parsed as a bool value

GustavoCaso commented on pkg/config/structure/unmarshal.go:
```diff
@@ -0,0 +1,368 @@
+}
+
+func newNode(v reflect.Value) node {
+	if v.Kind() == reflect.Struct {
+		return &nodeimpl{typ: structNodeType, val: v}
+	}
+	if v.Kind() == reflect.Map {
+		return &nodeimpl{typ: mapNodeType, val: v}
+	}
+	if v.Kind() == reflect.Slice {
+		return &nodeimpl{typ: listNodeType, val: v}
+	}
+	if isScalar(v) {
+		return &nodeimpl{typ: scalarNodeType, val: v}
+	}
```
Nitpick suggestion

```suggestion
	nodeType := invalidNodeType

	if v.Kind() == reflect.Struct {
		nodeType = structNodeType
	}
	if v.Kind() == reflect.Map {
		nodeType = mapNodeType
	}
	if v.Kind() == reflect.Slice {
		nodeType = listNodeType
	}
	if isScalar(v) {
		nodeType = scalarNodeType
	}

	if nodeType == invalidNodeType {
		return nil, fmt.Errorf("could not create node from: %v of type %T and kind %v", v, v, v.Kind())
	}

	return &nodeimpl{typ: nodeType, val: v}
```

dustmop replied:
Now this method returns a different implementation based upon the value's kind.

GustavoCaso commented on pkg/config/structure/unmarshal.go:
```diff
@@ -0,0 +1,368 @@
+	if isScalar(v) {
+		return &nodeimpl{typ: scalarNodeType, val: v}
+	}
+	panic(fmt.Errorf("could not create node from: %v of type %T and kind %v", v, v, v.Kind()))
```
Do we want to panic? Should we return `(node, error)` and handle the case of invalid node type?

dustmop replied:
Removed all panics from the file, returning errors everywhere now.

GustavoCaso commented on pkg/config/structure/unmarshal.go:
```diff
@@ -0,0 +1,368 @@
+	if outValue.Kind() == reflect.Pointer {
+		outValue = reflect.Indirect(outValue)
+	}
+	if outValue.Kind() == reflect.Map {
+		return copyMap(outValue, source)
+	}
+	if outValue.Kind() == reflect.Struct {
+		return copyStruct(outValue, source)
+	}
+	if outValue.Kind() == reflect.Slice {
+		return copyList(outValue, source.AsList())
+	}
```
Nitpick: Should we use a switch statement here?

dustmop replied:
Yes, good call, done.

GustavoCaso commented on pkg/config/structure/unmarshal.go:
```diff
@@ -0,0 +1,368 @@
+		}
+		scalar := child.AsScalar()
+		if scalar != nil {
+			mval, _ := scalar.GetString()
```
If understand correctly, `GetString` would return an empty string if leaf can not get the string value. Do we want to store empty string as map values?

dustmop replied:
Changed this call to `GetString` so that we check the error return. This call isn't quite a conversion: the var `scalar` represents a leaf node, and this call is getting the string value of that leaf node.

GustavoCaso commented on pkg/config/structure/unmarshal.go:
```diff
@@ -0,0 +1,368 @@
+	return false, conversionError(n.val, "bool")
+}
+
+// GetInt returns the scalar as a int, or an error otherwise
+func (n *leafimpl) GetInt() (int, error) {
+	switch n.val.Kind() {
+	case reflect.Int, reflect.Int8, reflect.Int16, reflect.Int32, reflect.Int64:
+		return int(n.val.Int()), nil
+	case reflect.Uint, reflect.Uint8, reflect.Uint16, reflect.Uint32, reflect.Uint64:
+		return int(n.val.Uint()), nil
+	}
+	return 0, conversionError(n.val, "int")
+}
```
YAML supports floats. Do we care about floats for our use case?

GustavoCaso replied:
We seem to use float values on our configuration template `pkg/config/config_template.yaml`

dustmop replied:
Added float support to the various places that needed it.

GustavoCaso commented on pkg/config/structure/unmarshal.go:
```diff
@@ -0,0 +1,368 @@
+	}
+	elemType := target.Type()
+	elemType = elemType.Elem()
+	results := reflect.MakeSlice(reflect.SliceOf(elemType), source.Size(), source.Size())
```
We are accessing `souorce.Size()` three times, should hosting the size value on a variable?

dustmop replied:
Sure, done, now `numElems` is used for this value.

GustavoCaso reviewed (commented):
Awesome work. The code is very easy to follow 🎉 

Left a few questions 😄

hush-hush commented on pkg/config/structure/unmarshal.go:
```diff
@@ -0,0 +1,368 @@
+	structNodeType
+)
+
+type leafimpl struct {
+	val reflect.Value
+}
+
+type leaf interface {
+	GetBool() (bool, error)
+	GetInt() (int, error)
+	GetString() (string, error)
+}
+
+type nodeListImpl struct {
+	val reflect.Value
+}
+
+type nodeList interface {
+	Size() int
+	Index(int) node
+}
+
+type nodeimpl struct {
+	typ nodeType
+	val reflect.Value
+}
+
+type node interface {
+	GetChild(string) node
+	ChildrenKeys() []string
+	AsList() nodeList
+	AsScalar() leaf
+}
```
Why have interface for each type here ? It seems we will only ever have 1 implementation for each one, no ?

dustmop replied:
This question made me realize a better approach for representing and constructing the nodes. Now each node is always the implementation that matches its type (inner node, leaf node, array node, map node) and `newNode` returns that implementation immediately, instead of carrying around the reflect kind (the `typ` member variable) to do so later. This removes the need for `AsScalar` and `AsList`, now we just use interface type assertions instead.

This plan is mirroring how the new config model will work. So the main idea is using interfaces instead of implementations in order to control how the model's internal data is being accessed. Getting type dispatch to handle node behavior is a nice benefit.

hush-hush commented on pkg/config/structure/unmarshal.go:
```diff
@@ -0,0 +1,368 @@
+		return n.val.Bool(), nil
+	}
+	if n.val.Kind() == reflect.String {
+		if n.val.String() == "true" {
+			return true, nil
+		} else if n.val.String() == "false" {
+			return false, nil
+		}
```
Sadly for things like yaml or json we need to support more aliasing as Go accept `true t yes y 1 on` when loading boolean.

dustmop replied:
Added the additional values to handle bools from yaml.

dustmop committed:
Merge branch 'main' into dustin.long/unmarshal-reflect

dustmop committed:
newNode decides on implmementation to use, support floats

Instead of methods like AsList and AsScalar, rely on interface type checking. Each specific type of node uses its own implementation.
Support struct tag specifiers like omitempty.
Introduce errNotFound for GetChild when nodes aren't found

dustmop commented on pkg/config/structure/unmarshal.go:
```diff
@@ -0,0 +1,368 @@
+		}
+		return keys
+	}
+	return nil
```
I was using `nil` to mean "no children" before, but now this function returns an error instead in that case.

dustmop commented on pkg/config/structure/unmarshal.go:
```diff
@@ -0,0 +1,368 @@
+	if outValue.Kind() == reflect.Slice {
+		return copyList(outValue, source.AsList())
+	}
+	return fmt.Errorf("can only UnmarshalKey to struct or slice, got %v", outValue.Kind())
```
Done.

dustmop commented on pkg/config/structure/unmarshal.go:
```diff
@@ -0,0 +1,368 @@
+func copyMap(target reflect.Value, source node) error {
+	// TODO: Should handle maps with more complex types
+	ktype := reflect.TypeOf("")
+	vtype := reflect.TypeOf("")
```
Yes, I think there's enough in this PR. Expanded on the TODO to specify that support for other maps will be in another PR.

dustmop commented on pkg/config/structure/unmarshal.go:
```diff
@@ -0,0 +1,368 @@
+		keyvals := n.val.MapKeys()
+		keys := make([]string, 0, len(keyvals))
+		for _, kv := range keyvals {
+			kinf := kv.Interface()
```
Oops I think I was spelling "interface" as "inf" which actually reads like "infinity". Regardless, the variable has been eliminated.

dustmop commented on pkg/config/structure/unmarshal.go:
```diff
@@ -0,0 +1,368 @@
+// ChildrenKeys returns the list of keys of the children of the given node
+func (n *nodeimpl) ChildrenKeys() []string {
+	if n.val.Kind() == reflect.Map {
+		keyvals := n.val.MapKeys()
```
Changed to `mapKeys`. Note that `vals` was being used because these are `reflect.Value` rather than normal strings.

dustmop commented on pkg/config/structure/unmarshal.go:
```diff
@@ -0,0 +1,368 @@
+			kstr, ok := kinf.(string)
+			if ok {
+				keys = append(keys, kstr)
+			}
```
Do we want / need to support general interface types for keys? It would be nice is the design of the config could move to using strings for keys everywhere, as that simplifies the design a bit and makes it easier to understand. In this specific case, the keys are returned here and then passed to `node.GetChild` which requires a string.

hush-hush replied:
As discussed we can start like this and expand if needed.

dustmop commented on pkg/config/structure/unmarshal.go:
```diff
@@ -0,0 +1,368 @@
+	if n.typ != mapNodeType && n.typ != structNodeType {
+		return nil
+	}
+	if n.val.Kind() == reflect.Map {
```
No longer applicable, as this decision is now handled by the implementation choice.

dustmop commented on pkg/config/structure/unmarshal.go:
```diff
@@ -0,0 +1,368 @@
+			name = replace
+		} else if replace := f.Tag.Get("mapstructure"); replace != "" {
+			name = replace
+		}
```
Good call, now a helper function is splitting the specifier from the struct tag.

dustmop commented on pkg/config/structure/unmarshal.go:
```diff
@@ -0,0 +1,368 @@
+			key = replace
+		} else if replace := f.Tag.Get("mapstructure"); replace != "" {
+			key = replace
+		}
```
Done. `fieldNameToKey` now exists.

dustmop committed:
Fix lint error about else return, more func docs

hush-hush commented on pkg/config/structure/unmarshal.go:
```diff
@@ -0,0 +1,440 @@
+// convert a string to a bool using standard yaml constants
+func convertToBool(text string) (bool, error) {
+	lower := strings.ToLower(text)
+	if lower == "y" || lower == "yes" || lower == "on" || lower == "true" {
+		return true, nil
+	} else if lower == "n" || lower == "no" || lower == "off" || lower == "false" {
+		return false, nil
+	}
```
The YAML/JSON go libs also support `1`/`0` for bools.

dustmop replied:
Ah makes sense, I'll add that before merging.

hush-hush commented on pkg/config/structure/unmarshal.go:
```diff
@@ -0,0 +1,440 @@
+	mtype := reflect.MapOf(ktype, vtype)
+	results := reflect.MakeMap(mtype)
+
+	mapKeys, err := source.ChildrenKeys()
```
This would mean we could Unmarshal a struct into a map, no ? Should we prevent this and check that the source is also a map ?

Similar question for the other `copy*` func

dustmop replied:
I think this is okay? It matches how the standard Unmarshal family of functions work. We'll keep an eye on it in future PRs, as this new function is used in more places, to see if we should decide to disable this behavior.

hush-hush reviewed (approved):
I added a couple of quesiton/nit-picks, but it looks good overall !0

dustmop committed:
Support bool values as 1 or 0 strings

dustmop commented:
/merge
